### PR TITLE
[NFC][CodeGen] Change CodeGenOpt::Level/CodeGenFileType into enum classes

### PR DESCRIPTION
**Pull request by aeubanks**
This will make it easy for callers to see issues with and fix up calls
to createTargetMachine after a future change to the params of
TargetMachine.

This matches other nearby enums.

For downstream users, this should be a fairly straightforward replacement,
e.g. s/CodeGenOpt::Aggressive/CodeGenOptLevel::Aggressive
or s/CGFT_/CodeGenFileType::

### CODE REVIEW

File: llvm/include/llvm/Support/CodeGen.h [hunk 1]
```diff
@@ -54,7 +54,7 @@ namespace llvm {
   /// Type for the unique integer IDs of code generation optimization levels.
   using IDType = int;
   /// Code generation optimization level.
-  enum Level : IDType {
+  enum class Level : IDType {
```
Comment (rnk) [inline llvm/include/llvm/Support/CodeGen.h:57]: Would you consider dropping the `namespace CodeGenOpt` and making this `enum class CodeGenOptLevel`? Then run `s/CodeGenOpt::Level/CodeGenOptLevel/` across the codebase. It is similar to `CodeGenFileType`. I would leave `parseLevel` and `getLevel` in their namespace.

**Review by joker-eph — APPROVED** [review]

**Review by rnk — COMMENTED** [review]
Seems reasonable to me.

File: llvm/include/llvm/Support/CodeGen.h (hunk 1 above)
Comment (aeubanks) [reply llvm/include/llvm/Support/CodeGen.h]: done

**Review by mtrofin — APPROVED** [review]

**Review by mshockwave — APPROVED** [review]

**Review by MaskRay — APPROVED** [review]

**Comment by ldionne** [comment]
There shouldn't be any changes under libcxxabi. Those changes are not necessary.

File: clang/lib/Frontend/CompilerInvocation.cpp [hunk 2]
```diff
@@ -661,27 +661,27 @@ static bool FixupInvocation(CompilerInvocation &Invocation,
 
 static unsigned getOptimizationLevel(ArgList &Args, InputKind IK,
                                      DiagnosticsEngine &Diags) {
-  unsigned DefaultOpt = llvm::CodeGenOpt::None;
+  unsigned DefaultOpt = 0;
   if ((IK.getLanguage() == Language::OpenCL ||
        IK.getLanguage() == Language::OpenCLCXX) &&
       !Args.hasArg(OPT_cl_opt_disable))
-    DefaultOpt = llvm::CodeGenOpt::Default;
+    DefaultOpt = 2;
```
[the remark this answers is not in the document]
Comment (banach-space) [reply clang/lib/Frontend/CompilerInvocation.cpp:668]: +1

Otherwise the the default is effectively defined twice in two different places.

**Review by banach-space — APPROVED** [review]
Flang driver changes LGTM, thanks!

File: clang/lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aeubanks) [reply clang/lib/Frontend/CompilerInvocation.cpp:668]: `llvm::CodeGenOpt` is for LLVM's codegen pipeline, but here it's getting overloaded to be used in clang's IR codegen. These are two different concepts where the defaults happen to have the same value.

**Review by rnk — APPROVED** [review]

File: clang/lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pogo59) [reply clang/lib/Frontend/CompilerInvocation.cpp:668]: Not that different; `CodeGenOpt::getLevel()` just uses a static_cast to implement the inverse mapping. This direction of the conversion should be consistent IMO.

[the remark this answers is not in the document]
Comment (aeubanks) [reply clang/lib/Frontend/CompilerInvocation.cpp:668]: We do eventually map the opt level here to a `CodeGenOpt` via `CodeGenOpt::getLevel()`, but we also use the opt level here for a bunch of other things in the clang codegen process. So whatever the "default" llvm codegen pipeline opt level is shouldn't be tied to a clang frontend opt level default.

Maybe I'm not understanding your comment.

[the remark this answers is not in the document]
Comment (pogo59) [reply clang/lib/Frontend/CompilerInvocation.cpp:668]: Actually it looks to me like clang only cares about 0/not-0, and the specific non-zero number is irrelevant. So, internally clang codegen would be better off with a simple Optimize bool. (It looks like LangOpts does it this way already.) CodeGenOpts probably does need to track the actual user-specified level to pass down to LLVM, but it doesn't need it for its own purposes.

**Pull request merged**